### DISCUSSION TITLE
Training with small vocab size or small context length causes reported training error to be cut by a factor of ~10

### DISCUSSION THREAD

**Discussion by dcymuller** [karpathy/llm.c#861] [Q&A] [answered]
I am training versions of this model with a reduced tokenizer vocabulary size and a reduced context length. When vocab size is reduced, I am using V=81 and Vp=128; when context length is reduced I am using T=4. When I am only reducing vocab size, the reported validation loss looks good (starts at around 4.5, which is around ln(81)), but the reported training loss starts at around 0.5. When I additionally reduce the context length, the reported training loss starts at 0.002. In all cases, both training and validation loss curves have the right pattern, and sampling from the model at the end of training shows the model has learned something. So it definitely seems like the issue is with the reporting of the training loss (and only the training loss!)— training of the model is not affected. Any ideas why this might be happening? I spent a bit of time tracing how the training loss is reported and how it is different from the validation loss, but I didn’t find any obvious/promising lines of code where both the vocab size and the context length contribute to the training loss computation.

I am training on one H100 PCIe GPU on Lambda Labs. 

Thank you in advance for any ideas!

**Comment by Yigtwxx** [comment, marked as answer]
Your two numbers pin this down better than the code does. Both of them are `ln(V) / grad_accum_steps`.

`ln(81) = 4.394`, which is what your validation loss reports. Against that, the reported *training* loss:

- V=81, T=1024 → ~0.5, so the divisor is ~8.8
- V=81, T=4 → ~0.002, so the divisor is ~2200

The ratio between those two divisors is ~250 and you changed T by a factor of 256. That's not a coincidence, because `grad_accum_steps` is derived from T:

```c
int tokens_per_fwdbwd = B * T * multi_gpu_config.num_processes;
if (total_batch_size == -1) { total_batch_size = tokens_per_fwdbwd; }
assert(total_batch_size % tokens_per_fwdbwd == 0);
int grad_accum_steps = total_batch_size / tokens_per_fwdbwd;
```

(`train_gpt2.cu:1513-1519`.) Hold `-d` fixed and cut T by 256x, and `grad_accum_steps` goes up by exactly 256x. So both data points fall out of one explanation: `grad_accum_steps` was 8 in the first run and 2048 in the second, and the reported training loss is the true loss divided by it. `4.394 / 8 = 0.549`. `4.394 / 2048 = 0.00215`. Vocab size isn't part of the mechanism at all — `ln(81)` just happens to be what the loss is.

This is a bug llm.c actually had. The reported loss is summed out of `acts.losses` and then divided by `B*T*grad_accum_steps`:

```c
global_sum_deterministic(model->accumulated_mean_loss, acts.losses, B*T, main_stream);
...
model->mean_loss /= B*T*grad_accum_steps;
```

(`train_gpt2.cu:955` and `:969`.) That divisor is only right if every micro-step accumulates into the buffer. The fused classifier used to overwrite instead:

```c
losses[idx] = (floatX)(-logf(prob));
```

so only the last micro-step survived and the reported loss came out as exactly `true_loss / grad_accum_steps`. Commit `8b36c2f` ("fix loss accumulation", 2024-06-24) changed it to accumulate, and moved the `acts.losses` memset under `micro_step == 0`, which is what master does now.

It also covers the part you couldn't find in the code. `dloss` is computed separately at `train_gpt2.cu:819` and is correct either way, so gradients and the actual optimization are untouched. And `gpt2_validate` runs its own forward with its own memset and no backward pass, so validation is untouched. Only the reported training number is wrong — which is the exact shape of what you're seeing.

So, two things to check.

Your run already prints `=> setting grad_accum_steps=N`. Multiply your reported training loss by that N. If you land near 4.4, that's the whole story.

Then check whether your checkout predates `8b36c2f`, or whether your own edits for the reduced vocab and context touched either the `losses[idx]` line in `fused_classifier_kernel5` or the `micro_step == 0` guard around the memset at `train_gpt2.cu:800`. If you're on unmodified master and `grad_accum_steps` really is 1, then I've read this wrong and it'd be worth posting B, T, your `-d` and the commit hash.

**Reply by dcymuller** [reply]
Ahhh thank you so much!! I forgot that I changed `fused_classifier_kernel5` to make it run with a decreased vocab size, and in the process it looks like the `if(threadIdx.x == 0)` guard got lost. Multiplying reported training loss by the `grad_accum_steps` recovers losses that are on a reasonable scale.

In case anyone else is trying to do runs with a decreased vocab size, here is the kernel function I used (with the updated fix to ensure correct reported training loss):

```template <bool WriteDLogits, bool WriteProbs>
__global__ void __launch_bounds__(1)
fused_classifier_kernel_small_vocab(
    floatX* logits,
    float* losses,
    floatX* probs,
    const float dloss,
    const int* targets,
    int B, int T,
    int V, int P,
    std::bool_constant<WriteDLogits>)
{
    // One block per row (same grid layout assumption)
    int idx = blockIdx.x;
    int N = B * T;
    if (idx >= N) return;

    int ix = targets[idx];
    floatX* row = logits + idx * P;

    // compute max
    float maxval = -INFINITY;
    for (int i = 0; i < V; i++) {
        maxval = fmaxf(maxval, (float)row[i]);
    }

    // compute denominator
    float sum = 0.0f;
    for (int i = 0; i < V; i++) {
        sum += expf((float)row[i] - maxval);
    }

    float scale = 1.0f / sum;

    // compute loss
    if(threadIdx.x == 0) {
        float prob = expf((float)row[ix] - maxval) * scale;
        losses[idx] = -logf(prob);
    }

    // backward
    for (int i = 0; i < V; i++) {
        float p = expf((float)row[i] - maxval) * scale;

        if constexpr (WriteDLogits) {
            float indicator = (i == ix) ? 1.0f : 0.0f;
            row[i] = (floatX)((p - indicator) * dloss);
        }

        if constexpr (WriteProbs) {
            if (probs != nullptr)
                probs[idx * P + i] = (floatX)p;
        }
    }
}```
